File: tests/parity/cases/commitment.ts
// Input cases for the commitment (commit-reveal ledger) parity vectors. This
// file runs ONLY at bless time: it may call the real impl to manufacture
// concrete, self-contained args (a ledger entry carries literal hashes), so the
// frozen vector needs no builder to replay - Dart feeds the same literal entries
// straight through its own verifyLedger.
import {
  COMMITMENT_GENESIS,
  computeCommitment,
  computeEntryHash,
  computeLeagueCommitment,
  computeLeagueEntryHash,
  computeSubject,
  type LeagueLedgerEntry,
  type LedgerEntry,
} from '#shared/commitment'

interface RawCase {
  fn: string
  args: unknown[]
}

const ISO1 = '2026-06-19T10:00:00.000Z'
const ISO2 = '2026-06-19T11:00:00.000Z'

async function makeEntry(
  seq: number,
  prevHash: string,
  userId: string,
  matchId: string,
  createdAt: string,
  opening: { homeGoals: number; awayGoals: number; salt: string },
  opened = true,
): Promise<LedgerEntry> {
  const subject = await computeSubject(userId)
  const commitment = await computeCommitment({ subject, matchId, ...opening })
  const entryHash = await computeEntryHash({ seq, prevHash, commitment, subject, matchId, createdAt })
  return { seq, prevHash, commitment, subject, matchId, createdAt, entryHash, opened, ...opening }
}

async function makeLeagueEntry(
  seq: number,
  prevHash: string,
  userId: string,
  leagueId: string,
  matchId: string,
  createdAt: string,
  opening: { homeGoals: number; awayGoals: number; salt: string },
  opened = true,
): Promise<LeagueLedgerEntry> {
  const subject = await computeSubject(userId)
  const commitment = await computeLeagueCommitment({ subject, leagueId, matchId, ...opening })
  const entryHash = await computeLeagueEntryHash({ seq, prevHash, commitment, subject, leagueId, matchId, createdAt })
  return { seq, prevHash, commitment, subject, leagueId, matchId, createdAt, entryHash, opened, ...opening }
}

export async function buildCases(): Promise<RawCase[]> {
  const cases: RawCase[] = []

  // --- primitives: string in, hex hash out ---
  cases.push({ fn: 'sha256Hex', args: [''] })
  cases.push({ fn: 'sha256Hex', args: ['abc'] })
  cases.push({ fn: 'sha256Hex', args: ['ngc-parity-éà🌐'] }) // non-ASCII: UTF-8 encoding must match
  cases.push({ fn: 'computeSubject', args: ['user-1'] })
  cases.push({ fn: 'computeSubject', args: ['user-2'] })
  cases.push({
    fn: 'computeCommitment',
    args: [{ subject: 's', matchId: 'm', homeGoals: 2, awayGoals: 1, salt: 'pepper' }],
  })
  cases.push({
    fn: 'computeEntryHash',
    args: [{ seq: 1, prevHash: COMMITMENT_GENESIS, commitment: 'c', subject: 's', matchId: 'm', createdAt: ISO1 }],
  })
  cases.push({
    fn: 'computeLeagueCommitment',
    args: [{ subject: 's', leagueId: 'L1', matchId: 'm', homeGoals: 2, awayGoals: 1, salt: 'pepper' }],
  })
  cases.push({
    fn: 'computeLeagueEntryHash',
    args: [{ seq: 1, prevHash: COMMITMENT_GENESIS, commitment: 'c', subject: 's', leagueId: 'L1', matchId: 'm', createdAt: ISO1 }],
  })

  // --- verifyLedger: the state machine, one case per outcome/reason ---
  const e1 = await makeEntry(1, COMMITMENT_GENESIS, 'u1', 'm1', ISO1, { homeGoals: 2, awayGoals: 1, salt: 'a' })
  const e2 = await makeEntry(2, e1.entryHash, 'u2', 'm2', ISO2, { homeGoals: 0, awayGoals: 0, salt: 'b' }, false)
  cases.push({ fn: 'verifyLedger', args: [[]] }) // empty slice -> genesis head
  cases.push({ fn: 'verifyLedger', args: [[e1, e2]] }) // valid, mixed opened/unopened
  cases.push({ fn: 'verifyLedger', args: [[{ ...e1, homeGoals: 5 }]] }) // bad opening -> reason: commitment
  cases.push({ fn: 'verifyLedger', args: [[{ ...e2, commitment: 'f'.repeat(64) }]] }) // forged hash -> entry-hash
  const gap = await makeEntry(3, e1.entryHash, 'u1', 'm3', ISO2, { homeGoals: 1, awayGoals: 0, salt: 'c' })
  cases.push({ fn: 'verifyLedger', args: [[e1, gap]] }) // seq gap -> reason: sequence
  const badLink = await makeEntry(2, 'wrong-prev', 'u1', 'm2', ISO2, { homeGoals: 1, awayGoals: 0, salt: 'b' })
  cases.push({ fn: 'verifyLedger', args: [[e1, badLink]] }) // broken link -> reason: link

  // --- witnessExtension: the consistency-proof state machine ---
  const w1 = await makeEntry(1, COMMITMENT_GENESIS, 'u1', 'm1', ISO1, { homeGoals: 1, awayGoals: 0, salt: 'a' })
  const w2 = await makeEntry(2, w1.entryHash, 'u1', 'm2', ISO2, { homeGoals: 2, awayGoals: 0, salt: 'b' })
  const w3 = await makeEntry(3, w2.entryHash, 'u1', 'm3', ISO1, { homeGoals: 0, awayGoals: 0, salt: 'c' })
  const w4 = await makeEntry(4, w3.entryHash, 'u1', 'm4', ISO2, { homeGoals: 1, awayGoals: 1, salt: 'd' })
  cases.push({ fn: 'witnessExtension', args: [null, [], { seq: 4, headHash: 'h' }] }) // first-seen
  cases.push({ fn: 'witnessExtension', args: [{ seq: 2, headHash: w2.entryHash }, [], { seq: 2, headHash: w2.entryHash }] }) // consistent
  cases.push({ fn: 'witnessExtension', args: [{ seq: 2, headHash: w2.entryHash }, [], { seq: 2, headHash: 'evil' }] }) // tampered
  cases.push({ fn: 'witnessExtension', args: [{ seq: 5, headHash: 'h5' }, [], { seq: 3, headHash: 'h3' }] }) // rolled-back
  cases.push({ fn: 'witnessExtension', args: [{ seq: 2, headHash: w2.entryHash }, [w3, w4], { seq: 4, headHash: w4.entryHash }] }) // valid extension
  // The three ways an extension is rejected (each OR-clause of the tampered
  // branch), so a Dart port can't pass by only proving the happy path:
  cases.push({ fn: 'witnessExtension', args: [{ seq: 2, headHash: w2.entryHash }, [w4], { seq: 4, headHash: w4.entryHash }] }) // non-contiguous (skips seq 3)
  cases.push({ fn: 'witnessExtension', args: [{ seq: 2, headHash: w2.entryHash }, [{ ...w3, prevHash: 'wrong-prev' }, w4], { seq: 4, headHash: w4.entryHash }] }) // broken link inside extension (verifyLedger fails)
  cases.push({ fn: 'witnessExtension', args: [{ seq: 2, headHash: w2.entryHash }, [w3, w4], { seq: 4, headHash: 'not-the-walked-head' }] }) // walks fine but does not reach the served head

  // --- league ledger: domain-separated twin ---
  const l1 = await makeLeagueEntry(1, COMMITMENT_GENESIS, 'u1', 'L1', 'm1', ISO1, { homeGoals: 2, awayGoals: 1, salt: 'a' })
  const l2 = await makeLeagueEntry(2, l1.entryHash, 'u2', 'L1', 'm2', ISO2, { homeGoals: 0, awayGoals: 0, salt: 'b' }, false)
  cases.push({ fn: 'verifyLeagueLedger', args: [[l1, l2]] })
  cases.push({ fn: 'verifyLeagueLedger', args: [[{ ...l1, homeGoals: 9 }]] })

  return cases
}

File: apps/mobile-flutter/parity/lib/commitment.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';

// Dart port of shared/commitment.ts (the commit-reveal prediction ledger).
// Operates on JSON-shaped maps/lists so a frozen vector replays through it with
// no model layer. Must stay byte-for-byte identical to the TS source: the
// parity vectors are the proof.

const genesis = '0000000000000000000000000000000000000000000000000000000000000000';
const _subjectPrefix = 'ngc-subject-v1:';
const _commitPrefix = 'ngc-commit-v1:';
const _chainPrefix = 'ngc-chain-v1:';
const _leagueCommitPrefix = 'ngc-lcommit-v1:';
const _leagueChainPrefix = 'ngc-lchain-v1:';

String sha256Hex(String input) => sha256.convert(utf8.encode(input)).toString();

String computeSubject(String userId) => sha256Hex(_subjectPrefix + userId);

String computeCommitment(Map o) =>
    sha256Hex(_commitPrefix + [o['subject'], o['matchId'], o['homeGoals'], o['awayGoals'], o['salt']].join(':'));

String computeEntryHash(Map l) =>
    sha256Hex(_chainPrefix + [l['seq'], l['prevHash'], l['commitment'], l['subject'], l['matchId'], l['createdAt']].join(':'));

String computeLeagueCommitment(Map o) => sha256Hex(
    _leagueCommitPrefix + [o['subject'], o['leagueId'], o['matchId'], o['homeGoals'], o['awayGoals'], o['salt']].join(':'));

String computeLeagueEntryHash(Map l) => sha256Hex(_leagueChainPrefix +
    [l['seq'], l['prevHash'], l['commitment'], l['subject'], l['leagueId'], l['matchId'], l['createdAt']].join(':'));

Map<String, dynamic> _fail(int count, String head, int failedSeq, String reason) =>
    {'ok': false, 'count': count, 'head': head, 'failedSeq': failedSeq, 'reason': reason};

Map<String, dynamic> verifyLedger(List entries, [String expectedPrev = genesis]) {
  var prev = expectedPrev;
  int? expectedSeq;
  for (final e in entries.cast<Map>()) {
    if (expectedSeq != null && e['seq'] != expectedSeq) return _fail(entries.length, prev, e['seq'], 'sequence');
    if (e['prevHash'] != prev) return _fail(entries.length, prev, e['seq'], 'link');
    if (e['opened'] == true) {
      final commitment = computeCommitment(e);
      if (commitment != e['commitment']) return _fail(entries.length, prev, e['seq'], 'commitment');
    }
    final entryHash = computeEntryHash(e);
    if (entryHash != e['entryHash']) return _fail(entries.length, prev, e['seq'], 'entry-hash');
    prev = e['entryHash'];
    expectedSeq = (e['seq'] as int) + 1;
  }
  return {'ok': true, 'count': entries.length, 'head': prev};
}

Map<String, dynamic> verifyLeagueLedger(List entries, [String expectedPrev = genesis]) {
  var prev = expectedPrev;
  int? expectedSeq;
  for (final e in entries.cast<Map>()) {
    if (expectedSeq != null && e['seq'] != expectedSeq) return _fail(entries.length, prev, e['seq'], 'sequence');
    if (e['prevHash'] != prev) return _fail(entries.length, prev, e['seq'], 'link');
    if (e['opened'] == true) {
      final commitment = computeLeagueCommitment(e);
      if (commitment != e['commitment']) return _fail(entries.length, prev, e['seq'], 'commitment');
    }
    final entryHash = computeLeagueEntryHash(e);
    if (entryHash != e['entryHash']) return _fail(entries.length, prev, e['seq'], 'entry-hash');
    prev = e['entryHash'];
    expectedSeq = (e['seq'] as int) + 1;
  }
  return {'ok': true, 'count': entries.length, 'head': prev};
}

Map<String, dynamic> witnessExtension(Map? pin, List extension, Map servedHead) {
  if (pin == null) return {'status': 'first-seen', 'head': servedHead};
  if (servedHead['seq'] < pin['seq']) return {'status': 'rolled-back', 'head': pin};
  if (servedHead['seq'] == pin['seq']) {
    return servedHead['headHash'] == pin['headHash']
        ? {'status': 'consistent', 'head': servedHead}
        : {'status': 'tampered', 'head': pin};
  }
  final res = verifyLedger(extension, pin['headHash']);
  final contiguous = extension.isNotEmpty && (extension.first as Map)['seq'] == pin['seq'] + 1;
  if (res['ok'] != true || !contiguous || res['head'] != servedHead['headHash']) {
    return {'status': 'tampered', 'head': pin};
  }
  return {'status': 'consistent', 'head': servedHead};
}
